cmd/virt-xml-hook/virt_xml_hook.go

```diff
@@ -0,0 +1,139 @@
+
+	"github.com/spf13/pflag"
+	"google.golang.org/grpc"
+
+	"kubevirt.io/client-go/log"
+	"kubevirt.io/kubevirt/pkg/hooks"
+	hooksInfo "kubevirt.io/kubevirt/pkg/hooks/info"
+	hooksV1alpha1 "kubevirt.io/kubevirt/pkg/hooks/v1alpha1"
+	hooksV1alpha2 "kubevirt.io/kubevirt/pkg/hooks/v1alpha2"
+)
+
+const (
+	onDefineDomainLoggingMessage = "Hook's OnDefineDomain callback method has been called"
+	usage                        = `updater
+  --version v1alpha1|v1alpha2
+  --args ''`
+)
+
+type infoServer struct {
+	Version string
+}
+
+func (s infoServer) Info(ctx context.Context, params *hooksInfo.InfoParams) (*hooksInfo.InfoResult, error) {
+	log.Log.Info("Hook's Info method has been called")
+
+	return &hooksInfo.InfoResult{
+		Name: "update-xml",
+		Versions: []string{
+			s.Version,
+		},
+		HookPoints: []*hooksInfo.HookPoint{
+			{
+				Name:     hooksInfo.OnDefineDomainHookPointName,
+				Priority: 0,
+			},
+		},
+	}, nil
+}
+
+const virtXML = "virt-xml"
+
+type v1alpha1Server struct {
+	args []string
+}
+
+type v1alpha2Server struct {
+	args []string
+}
+
+func MergeKubeVirtXMLWithProvidedXML(domainXML []byte, args []string) ([]byte, error) {
+	args = append(args, "--edit")
+	args = append(args, "--print-xml")
+	cmd := exec.Command(virtXML, args...)
+	cmd.Stdin = strings.NewReader(string(domainXML))
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+	log.Log.Infof("Execute command: %s", cmd.String())
+	out, err := cmd.Output()
+	if err != nil {
+		log.Log.Errorf("Fail running command stdout:%s stderr: %s error:%v", out, stderr.String(), err)
+		return []byte{}, err
+	}
+
+	return []byte(out), nil
+}
+
+func (s v1alpha2Server) OnDefineDomain(ctx context.Context, params *hooksV1alpha2.OnDefineDomainParams) (*hooksV1alpha2.OnDefineDomainResult, error) {
+	log.Log.Info(onDefineDomainLoggingMessage)
+	newDomainXML, err := MergeKubeVirtXMLWithProvidedXML(params.GetDomainXML(), s.args)
+	if err != nil {
+		return nil, err
+	}
+	return &hooksV1alpha2.OnDefineDomainResult{
+		DomainXML: newDomainXML,
+	}, nil
+}
+func (s v1alpha2Server) PreCloudInitIso(_ context.Context, params *hooksV1alpha2.PreCloudInitIsoParams) (*hooksV1alpha2.PreCloudInitIsoResult, error) {
+	return &hooksV1alpha2.PreCloudInitIsoResult{
+		CloudInitData: params.GetCloudInitData(),
+	}, nil
+}
+
+func (s v1alpha1Server) OnDefineDomain(ctx context.Context, params *hooksV1alpha1.OnDefineDomainParams) (*hooksV1alpha1.OnDefineDomainResult, error) {
+	log.Log.Info(onDefineDomainLoggingMessage)
+	newDomainXML, err := MergeKubeVirtXMLWithProvidedXML(params.GetVmi(), s.args)
+	if err != nil {
+		return nil, err
+	}
+	return &hooksV1alpha1.OnDefineDomainResult{
+		DomainXML: newDomainXML,
+	}, nil
+}
+
+func main() {
+	log.InitializeLogging("xml update")
+
+	var version string
+	pflag.StringVar(&version, "version", "", "hook version to use")
+
+	var options string
+	pflag.StringVar(&options, "args", "", "params to pass to virt-xml")
+	pflag.Parse()
+
+	args := strings.Split(options, "|")
```

Comment:
Using pipe to separate arguments is kinda yucky. Can we do better? A few ideas.

We could use space as the separator, as in

```yaml
hooks.kubevirt.io/hookSidecars: '[{"args": ["--version", "v1alpha2", "--args", "--disk=iotune.total_bytes_sec=52428800,iotune.total_iops_sec=1000 --edit=all"], "image": "registry:5000/kubevirt/virt-xml-hook:devel"}]'
```

Would require escaping any whitespace that needs to be passed verbatim to virt-xml, which I'm not sure would work.

We could make the argument to `--args` a JSON array, as in

```yaml
hooks.kubevirt.io/hookSidecars: '[{"args": ["--version", "v1alpha2", "--args", "[\"--disk=iotune.total_bytes_sec=52428800,iotune.total_iops_sec=1000\", \"--edit=all\"]"], "image": "registry:5000/kubevirt/virt-xml-hook:devel"}]'
```

Assuming the extra quoting would do the trick, it would probably be annoying.

We could allow `--args` to be specified multiple times and accumulate the values, as in

```yaml
hooks.kubevirt.io/hookSidecars: '[{"args": ["--version", "v1alpha2", "--args", "--disk=iotune.total_bytes_sec=52428800,iotune.total_iops_sec=1000", "--args", "--edit=all"], "image": "registry:5000/kubevirt/virt-xml-hook:devel"}]'
```

It'd be more verbose, and I'm not sure pflag actually supports this.

We could expect virt-xml args as rest arguments, as in

```yaml
hooks.kubevirt.io/hookSidecars: '[{"args": ["--version", "v1alpha2", "--", "--disk=iotune.total_bytes_sec=52428800,iotune.total_iops_sec=1000", "--edit=all"], "image": "registry:5000/kubevirt/virt-xml-hook:devel"}]'
```

Would require the user to pass `--` before the first virt-xml option, which becomes especially awkward if we can do away with the `--version` option as per my other comment.

We could ask the user to provide args as "rest arguments" and without the leading `--`, as in

```yaml
hooks.kubevirt.io/hookSidecars: '[{"args": ["--version", "v1alpha2", "disk=iotune.total_bytes_sec=52428800,iotune.total_iops_sec=1000", "edit=all"], "image": "registry:5000/kubevirt/virt-xml-hook:devel"}]'
```

Might look surprising to people who are already used to virt-xml.

Reply:
Not sure. All the options are ugly :). I actually tried with the space separator but the second flag was interpreted as domain name so there is some parsing issue.